website/blog/2023-07-14-Local-LLMs/index.mdx

```diff
@@ -0,0 +1,134 @@
+---
+title: Use flaml.autogen for local LLMs
+authors: jialeliu
+tags: [LLM, FLAMLv2]
+---
+**TL;DR:**
+We demonstrate how to use flaml.autogen for local LLM application. As an example, we will initiate an endpoint using [FastChat](https://github.com/lm-sys/FastChat) and perform inference on [ChatGLMv2-6b](https://github.com/THUDM/ChatGLM2-6B).
+
+## Preparations
+
+### Clone FastChat
+
+FastChat provides OpenAI-compatible APIs for its supported models, so you can use FastChat as a local drop-in replacement for OpenAI APIs. However, its code needs minor modification in order to function properly.
+
+```bash
+git clone https://github.com/lm-sys/FastChat.git
+cd FastChat
+```
+
+According to this [issue](https://github.com/lm-sys/FastChat/issues/1641), comment out all the lines containing `finish_reason` in `fastchat/protocol/api_protocal.py` and `fastchat/protocol/openai_api_protocol.py` will fix the errors. The modified code looks like:
+
+```python
+class CompletionResponseChoice(BaseModel):
+    index: int
+    text: str
+    logprobs: Optional[int] = None
+    # finish_reason: Optional[Literal["stop", "length"]]
+
+class CompletionResponseStreamChoice(BaseModel):
+    index: int
+    text: str
+    logprobs: Optional[float] = None
+    # finish_reason: Optional[Literal["stop", "length"]] = None
+```
+
+### Download checkpoint
+
+ChatGLM-6B is an open bilingual language model based on General Language Model (GLM) framework, with 6.2 billion parameters. ChatGLM2-6B is its second-generation version.
+
+Before downloading from HuggingFace Hub, you need to have Git LFS [installed](https://docs.github.com/en/repositories/working-with-files/managing-large-files/installing-git-large-file-storage).
+
+```bash
+git clone https://huggingface.co/THUDM/chatglm2-6b
+```
+
+## Initiate server
+
+First, launch the controller
+
+```bash
+python -m fastchat.serve.controller
+```
+
+Then, launch the model worker(s)
+
+```bash
+python -m fastchat.serve.model_worker --model-path chatglm2-6b
+```
+
+Finally, launch the RESTful API server
+
+```bash
+python -m fastchat.serve.openai_api_server --host localhost --port 8000
+```
+
+## Interact with model using `oai.Completion`
+
+Now the models can be directly accessed through openai-python library as well as `flaml.oai.Completion`. If you are interacting with a chat model, make sure you have added your model name to `oai.Completion.chat_models`.
```

Comment:
It'll be convenient to remove the need for users to modify the code. We want them to be able to user our library without modifying it. Could you update the instructions?

Reply:
I'm afraid this is a prerequisite for `Completion` to interact with chat model. It requires `openai_completion` to become `openai.ChatCompletion` [here](https://github.com/microsoft/FLAML/blob/297a1ea9e0ef2a1a4bf0fa903ff63b2d46838944/flaml/autogen/oai/completion.py#L193) in order for chat completion function to work properly. I'm thinking of adding a `register_chat_model` to `Completion` for users to add the name of their local LLMs in the future.

Reply:
`if config["model"] in cls.chat_models or issubclass(cls, ChatCompletion)`
It works if we use `oai.ChatCompletion.create`.
`register_chat_model` is a good idea too.

Reply:
True, I didn't think of using `oai.ChatCompletion`. I will update the docs.